732 general update localization documentation (#786)
una-auxme/paf: doc/localization/coordinate_transformation.md

# Coordinate Transformation

**Summary:** Used for various helper functions such as quat_to_heading, that can be useful in various cases.

**Disclaimer:** It is not yet fully documented.

> Only the quat_to_heading function is, because it is the only function currently used.
> 

Reviewer: ```suggestion
Only the quat_to_heading and geodetic_to_enu(gnss_to_xyz) functions are, because these are the only functions currently used.

```

- [Usage](#usage)
- [Overview of the code](#overview-of-the-code)

>   - [The quat_to_heading function](#the-quat_to_heading-function)

Reviewer: ```suggestion
  - [The quat\_to\_heading function](#the-quat_to_heading-function)
  - [geodetic\_to\_enu (gnss\_to\_xyz)](#geodetic_to_enu-gnss_to_xyz)
```

Therefore, if the car is not deflected, its position can be described with the following unit vector:

$$
V =
\begin{bmatrix}
    1\\
    0\\
    0\\
\end{bmatrix}
$$

To calculate a rotated vector `V'`, the following calculation is performed:

$$
V' = R \cdot V
$$

$$
\begin{bmatrix}
    a\\
    d\\
    g\\
\end{bmatrix}
\quad = \quad
\begin{bmatrix}
    a & b & c\\
    d & e & f\\
    g & h & i\\
\end{bmatrix}
\cdot
\begin{bmatrix}
    1\\
    0\\
    0\\
\end{bmatrix}
$$

The result is a vector, that is rotated in the x-y-plane with the new x and y coordinates being `a` and `d`:

![Visualization quat_to_heading](../assets/localization/quat_to_heading.png)

Now the angle $\theta$ around the z-axis, which correlates to the heading, can be calculated using the `atan2` function:

$$
\theta = atan2(\frac{d}{a})
$$

However, Carla does not use a right-handed coordinate system like it is normally used in mathematics and in the image above.
It uses a left-handed coordinate system instead.
To incorporate that into the calculation, $\theta$ simply needs to be negated to get the correct heading.

$$
\text{heading} = - \theta
$$

> 

Reviewer: ```suggestion
### geodetic_to_enu (gnss_to_xyz)

This method is used by the [gps_transform](../../code/localization/src/gps_transform.py) node. The Carla GNSS-Sensor provides a lattitude, longitude and altitude measurement.
With the `geodetic_to_enu/gnss_to_xyz` method this can be transformed to x, y, z space. As used in all other systems of our vehicle.

The `CoordinateTransformer` class is built in a way that it is possible to provide a reference lattitude and longitude (which lat, lon equals (0,0) in x-y space). This reference is provided in the OpenDrive map received from Carla at startup. In the current scenarios this reference is always 0, 0.
If newer versions/scenarios change this reference adjustments might need to be made.

```

### Other available functions

- ecef_to_enu
- geodetic_to_ecef

> - geodetic_to_enu

Reviewer: ```suggestion

```
